### PR TITLE
feat: persistent websocket [AR-1168]

### PR DESCRIPTION
**Pull request by MohammedMokresh**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

we need to save a value in our storage to know if the user enable or disabled the websocket

### Solutions

create a simple SP value to save that flag 

Needs releases with:

- [ ] GitHub link to other pull request

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+interface EnableWebSocketUseCase {
+    operator fun invoke(enabled: Boolean): Either<StorageFailure, Unit>
+}
+
+class EnableWebSocketUseCaseImpl(
+    private val userConfigRepository: UserConfigRepository
+) : EnableWebSocketUseCase {
+    override operator fun invoke(enabled: Boolean) = userConfigRepository.persistWebSocketStatus(enabled)
+}
```
Comment (alexandreferris) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt:15]: If we have its name as `Enable` then we should not receive a boolean and just assume we are passing a boolean true to `persistWebSocketStatus`, no? 🤔

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.wire.kalium.logic.feature.user.webSocketStatus
+
+import com.wire.kalium.logic.StorageFailure
+import com.wire.kalium.logic.configuration.UserConfigRepository
+import com.wire.kalium.logic.functional.Either
+
+interface EnableWebSocketUseCase {
+    operator fun invoke(enabled: Boolean): Either<StorageFailure, Unit>
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt:8]: We should not expose `Either` to clients, so I think we need to wrap this in a `Result` class

**Review by yamilmedina — COMMENTED** [review]
Pretty good job, just some comments about class visibility and as well if we add more tests, we could have some smiley cats (😸)

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt (hunk 1 above)
Comment (MohammedMokresh) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/webSocketStatus/EnableWebSocketUseCase.kt:15]: hmmm  , you have a point there , I think it's better to go with the name persist web socket  status

**Review by alexandreferris — APPROVED** [review]
👌🏻

**Review by yamilmedina — APPROVED** [review]
😸

**Review by vitorhugods — CHANGES_REQUESTED** [review]
The naming is a bit weird.
Websocket is enabled. There's no way to disable it. We need it for sync.

The fact that it is `false` by default is very misleading.

Now... having a PERSISTENT Websocket connection is a different thing. So I think it makes more sense to have a clear distinction in the naming.

----

I don't quite get it.
This is an Android-specific issue. More than that... it's an F-Droid specific issue. No?

Why is this being implemented in Kalium? Would we add dark theme preference settings to Kalium too, for example? Where to draw the line?

I feel like this is a client-specific setting and that this should be implemented in Reloaded directly. If a user preference is on, we launch a service to keep the app alive in the background and set the `ConnectionPolicy` to `KEEP_ALIVE`.

**Review by vitorhugods — APPROVED** [review]
As we talked in private, changing this to whatever different approach now or in the future will have pretty much the same cost.
We need to discuss and bring forward solutions to having platform-specific UseCases and/or exposing a multi-platform Preferences solution so the apps can use it and create their own user-scoped settings.

**Pull request merged**